### PR TITLE
Run Lua interactive commands in Lua Task

### PR DESCRIPTION
**Pull request by stieg**
This semi-complex change causes us to run interactive Lua commands
in the Lua Task as opposed to the task hosting the interactive Lua
session.  This is done because we don't know the stack depth of
the task hosting the interactive session, hence we can't gaurantee
it will not overflow when invoking Lua commands.

Issue: #844

### CODE REVIEW

**Comment by stieg** [comment]
Am going to do a regression test of Lua tomorrow but basic testing both with and without a script showed that the original issue reported is now fixed.  Just need to ensure that no regressions were introduced by this.

File: src/lua/luaCommands.c [hunk 1]
```diff
@@ -39,14 +40,13 @@ void ExecLuaInterpreter(struct Serial *serial, unsigned int argc, char **argv)
                      "to leave");
         put_crlf(serial);
 
-        cmd_context *cmdContext = get_command_context();
-        char *luaLine = cmdContext->lineBuffer;
+        static char luaLine[256];
```
Comment (brentpicasso) [inline src/lua/luaCommands.c]: meh, lots of space in the CCM, right? :)

File: src/lua/luaTask.c [hunk 2]
```diff
@@ -321,42 +376,67 @@ static lua_State* setup_lua_state()
...
+	 * Notice we give the lua_signal twice here. The first time will
+	 * unblock the luaTask so that it may perform our command.  The
+	 * second call completes the interactive command.  This is done this
+	 * way so we may pop all elements/error messages as needed. Therefore
+	 * it is safe to manipulate the Lua state since the lua_interactive
+	 * method will keep the Lua lock until we give the lua_signal the
+	 * second time.
+	 *
+	 * Once this task has completed reading the results of the command
+	 * and has given the lua_signa, the lua_interactive task will
```
Comment (brentpicasso) [inline src/lua/luaTask.c]: spelling: lua_signal

File: src/lua/luaTask.c [hunk 3]
```diff
@@ -227,9 +275,16 @@ static void lua_task(void *params)
         };
 
         int consecutive_failures = 0;
-        for(portTickType xLastWakeTime;;
-            vTaskDelayUntil(&xLastWakeTime, state.callback_interval)) {
-                xLastWakeTime = xTaskGetTickCount();
+	portTickType wake_tick = 0;
+        for(;;) {
```
Comment (brentpicasso) [inline src/lua/luaTask.c]: so.. how does this play into how responsive the lua command is based on the current tickRate()?

Will a user enter a command, and then wait for the current tick to fire to see it execute? e.g. default tick rate is 1Hz - it could take up to 1 second for a command to be processed?

**Comment by brentpicasso** [comment]
wow that's a lot but it's good. the thread interaction can be tricky; we should both hammer on this tomorrow. 

Once main question in the changeset.

**Comment by brentpicasso** [comment]
detected bug:

if onTick() is not defined, I get an error:

RaceCapture/Pro MK2 > lua                                                                                                                                  
Entering Lua Interpreter. enter 'exit' to leave                                                                                                            
> println('hello')                                                                                                                                         
Internal Error: Lua not active.                                                                                                                            
>

File: src/lua/luaTask.c (hunk 3 above)
Comment (brentpicasso) [inline src/lua/luaTask.c]: Testing it directly it seems the command is processing immediately, even if tickRate() is 1Hz, which seems good, but it seems there's something else I'm not understanding out of reviewing the code.

Comment (brentpicasso) [inline src/lua/luaTask.c]: More data. while at 1Hz, I rapidly pasted the following into interactive mode, with a carriage return in the clipboard, so it processed immediately after the paste. 

println('asdfasdf')


The commands were happening quickly with manual pastes, but if I held down the keyboard to have it repeat, then the commands started processing at 1Hz! interesting.

File: src/lua/luaCommands.c (hunk 1 above)
Comment (stieg) [inline src/lua/luaCommands.c]: Yep.  I have a ticket out to be able to pass in the object with the buffer to handle this.. just not part of this patch and not a requirement for release.  Hence this hack.

**Comment by stieg** [comment]
Its expected, but lets talk about it since I'm not sure if it's a bug or a feature.

**Comment by brentpicasso** [comment]
Tested the latest. :shipit:  :D

**Comment by stieg** [comment]
You are up early @brentpicasso .  Merging.
